cmd/restic/cmd_serve.go

```diff
@@ -0,0 +1,268 @@
+			rows = append(rows, indexPageRow{"/tree/" + sn.ID().Str() + "/", sn.ID().Str(), sn.Time, sn.Hostname, sn.Tags, sn.Paths})
+		}
+		sort.Slice(rows, func(i, j int) bool {
+			return rows[i].Time.After(rows[j].Time)
+		})
+		if err := indexPage.Execute(w, indexPageData{"Snapshots", rows}); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	})
+
+	http.HandleFunc("/style.css", func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Cache-Control", "max-age=300")
+		_, _ = w.Write([]byte(stylesheetTxt))
+	})
+
+	Printf("Now serving the repository at http://%s\n", opts.Listen)
+	Printf("When finished, quit with Ctrl-c here.\n")
+
+	return http.ListenAndServe(opts.Listen, nil)
+}
+
+type indexPageRow struct {
+	Link  string
+	ID    string
+	Time  time.Time
+	Host  string
+	Tags  []string
+	Paths []string
+}
+
+type indexPageData struct {
+	Title string
+	Rows  []indexPageRow
+}
+
+type treePageRow struct {
+	Link string
+	Name string
+	Type string
+	Size uint64
+	Time time.Time
+}
+
+type treePageData struct {
+	Title  string
+	Parent string
+	Rows   []treePageRow
+}
+
+const indexPageTpl = `<html>
+<head>
+<link rel="stylesheet" href="/style.css">
+<title>{{.Title}} :: restic</title>
+</head>
+<body>
+<h1>{{.Title}}</h1>
+<table>
+<thead><tr><th>ID</th><th>Time</th><th>Host</th><th>Tags</th><th>Paths</th></tr></thead>
+<tbody>
+{{range .Rows}}
+<tr><td><a href="{{.Link}}">{{.ID}}</a></td><td>{{.Time | FormatTime}}</td><td>{{.Host}}</td><td>{{.Tags}}</td><td>{{.Paths}}</td></tr>
+{{end}}
+</tbody>
+</table>
+</body>
+</html>`
+
+const treePageTpl = `<html>
+<head>
+<link rel="stylesheet" href="/style.css">
+<title>{{.Title}} :: restic</title>
+</head>
+<body>
+<h1>{{.Title}}</h1>
+<form method="post">
+<table>
+<thead><tr><th><input type="checkbox" onclick="document.querySelectorAll('.content input[type=checkbox]').forEach(cb => cb.checked = this.checked)"></th><th>Name</th><th>Type</th><th>Size</th><th>Date modified</th></tr></thead>
+<tbody class="content">
+{{if .Parent}}<tr><td></td><td><a href="{{.Parent}}">..</a></td><td>parent</td><td></td><td></tr>{{end}}
+{{range .Rows}}
+<tr><td><input type="checkbox" name="name" value="{{.Name}}"></td><td><a class="{{.Type}}" href="{{.Link}}">{{.Name}}</a></td><td>{{.Type}}</td><td>{{.Size}}</td><td>{{.Time | FormatTime}}</td></td></tr>
+{{end}}
+</tbody>
+<tbody class="actions">
+<tr><td colspan="100"><button name="action" value="dump" type="submit">Download selection</button></td></tr>
+</tbody>
+</table>
+</form>
+</body>
+</html>`
+
+const stylesheetTxt = `
+h1,h2,h3 {text-align:center; margin: 0.5em;}
+table {margin: 0 auto;border-collapse: collapse; }
+thead th {text-align: left; font-weight: bold;}
+tbody.content tr:hover {background: #eee;}
+tbody.content a.file:before {content: '\1F4C4'}
+tbody.content a.dir:before {content: '\1F4C1'}
+tbody.actions td {padding:.5em;}
+table, td, tr, th { border: 1px solid black; padding: .1em .5em;}
+`
```

Comment:
May I suggest to use the Go embedding (`go:embed`) for these files?

Reply:
I considered embedding but it would mean finding the correct place to store assets and adding the extra build step for bundling and I assumed (perhaps incorrectly?) that if I did that, it would reduce the chances of this PR being accepted.

So in other words I agree on the principle, I'm just not sure I'm the best person to figure out all the details :).

Reply:
Actually, Go embedding does not require extra steps: https://pkg.go.dev/embed

Given that these assets need to be present during the build only, I think that a sub-directory (e.g. `serve-assets/` ?) may be OK. Let's see what restic developers think about this solution :-)